consensus/src/consensus/mod.rs

```diff
@@ -696,6 +700,120 @@ impl ConsensusApi for Consensus {
+                    return ReturnAddress::NotFound(format!("Did not find a compact header with hash {}", hash));
+                }
+            };
+
+            // 3. Compare block daa score to our target
+            match compact_header.daa_score.cmp(&target_daa_score) {
+                cmp::Ordering::Equal => {
+                    // We found the chain block we need
+                    break hash;
+                }
+                cmp::Ordering::Greater => {
+                    high_index = mid - 1;
+                }
+                cmp::Ordering::Less => {
+                    low_index = mid + 1;
+                }
+            }
+
+            if low_index > high_index {
+                return ReturnAddress::NoTxAtScore;
+            }
+        };
+
+        let acceptance_data = match self.acceptance_data_store.get(matching_chain_block_hash) {
+            Ok(acceptance_data) => acceptance_data,
+            Err(_) => {
+                return ReturnAddress::NotFound("Did not find acceptance data".to_string());
+            }
+        };
+        let (index, containing_acceptance) = match acceptance_data.iter().find_map(|mbad| {
+            let tx_arr_index =
+                mbad.accepted_transactions.iter().find_map(|tx| (tx.transaction_id == txid).then_some(tx.index_within_block as usize));
+            tx_arr_index.map(|index| (index, mbad.clone()))
+        }) {
+            Some((index, containing_acceptance)) => (index, containing_acceptance),
+            None => {
+                return ReturnAddress::NotFound("Did not find containing_acceptance".to_string());
+            }
+        };
+
+        // Found Merged block containing the TXID
+        let tx = &self.block_transactions_store.get(containing_acceptance.block_hash).unwrap()[index];
+
+        if tx.id() != txid {
+            // Should never happen, but do a sanity check. This would mean something went wrong with storing block transactions
+            // Sanity check is necessary to guarantee that this function will never give back a wrong address (err on the side of NotFound)
+            warn!("Expected {} to match {} when checking block_transaction_store using array index of transaction", tx.id(), txid);
+            return ReturnAddress::NotFound(format!(
+                "Expected {} to match {} when checking block_transaction_store using array index of transaction",
+                tx.id(),
+                txid
+            ));
+        }
+
+        if tx.inputs.is_empty() {
+            // A transaction may have no inputs (like a coinbase transaction)
+            return ReturnAddress::TxFromCoinbase;
+        }
+
+        let first_input_prev_outpoint = &tx.inputs[0].previous_outpoint;
+        // Expected to never fail, since we found the acceptance data and therefore there must be matching diff
+        let utxo_diff = self.utxo_diffs_store.get(matching_chain_block_hash).unwrap();
```

Comment:
There might be a rare case where the tx spends an output of another tx which was accepted by this same chain block (contained within another merged block which proceeded it in the mergeset ordering). In which case the UTXO diff will not contain the created-and-immediately-spent UTXO.  The solution in that case is to get the UTXO address from the output of that tx which must appear in this chain block's acceptance data as well

Reply:
Actually the comment should belong on the line `removed_diffs.get(first_input_prev_outpoint).unwrap()` below

Reply:
> The solution in that case is to get the UTXO address from the output of that tx which must appear in this chain block's acceptance data as well

Let me check my understanding

Scenario:
- UTXO0 is spent by TX1 and creates UTXO1
- UTXO1 is spent by TX2 and creates UTXO2
- A chain block happens to accept both of these

In this scenario, I know the details of UTXO2 and it's my input to the function. Because both txs are accepted by the same chain block, `removed_diff` will not contain the details of UTXO1 (corresponds to `first_input_prev_outpoint`). So I need to find TX1 from `block_transactions_store` instead and look at the output that corresponds to UTXO1. Is this accurate?

Reply:
one more comment: be defensive here and avoid unwrapping if possible (just in case we are missing more edge-case details like this). So prefer returning an RPC err + warn log with max details

Reply:
The edge case has been addressed. Remaining to do is refactor further and handle unwraps safely

Reply:
Refactored and handled all the unwraps. I changed the return of this function to be a Result, instead of just the direct enum.